objects/c932e8/script.lua

```diff
@@ -13,42 +44,161 @@ function onLoad()
     end
 end
 
+function onSave()
+    return JSON.encode{
+        initialized=initialized,
+        serpent=serpent and serpent.guid,
+        elementScanZone=elementScanZone,
+    }
+end
+
 function setupDeepSlumber(_, color)
     doSpiritSetup{color=color}
 end
 
+function findSerpent(zone)
+    local objs = getObjectFromGUID(zone).getObjects()
+    for _,obj in pairs(objs) do
+        if obj.guid == "cebe09" then
+            return obj
+       end
+    end
+    return nil
+end
+
+function findColor(zoneGUID)
+    for color, guid in pairs(Global.getVar("elementScanZones")) do
+        if guid == zoneGUID then
+            return color
+        end
+    end
+    return nil
+end
+
 function doSpiritSetup(params)
     local color = params.color
     if not Global.getVar("gameStarted") then
         Player[color].broadcast("Please wait for the game to start before pressing button!", "Red")
         return
     end
 
-    local zone = getObjectFromGUID(Global.getVar("elementScanZones")[color])
-    local objs = zone.getObjects()
-    local found = false
-    local serpent = nil
-    for _,obj in pairs(objs) do
-        if obj.guid == "cebe09" then
-            found = true
-            serpent = obj
-            break
-       end
-    end
-    if not found then
+    elementScanZone = Global.getVar("elementScanZones")[color]
+
+    serpent = findSerpent(elementScanZone)
+    if serpent == nil then
         Player[color].broadcast("You have not picked Serpent Slumbering Beneath the Island!", "Red")
         return
     end
 
     local snapPoints = serpent.getSnapPoints()
-
-    table.insert(snapPoints,{position={1.14, 0.20, 0.65}})
-    table.insert(snapPoints,{position={0.92, 0.20, 0.65}})
-    table.insert(snapPoints,{position={0.70, 0.20, 0.65}})
-    table.insert(snapPoints,{position={1.25, 0.20, 0.84}})
-    table.insert(snapPoints,{position={1.03, 0.20, 0.84}})
-    table.insert(snapPoints,{position={0.81, 0.20, 0.84}})
+    for _, position in pairs(slumberPoints) do
+        table.insert(snapPoints,{position=position})
+    end
     serpent.setSnapPoints(snapPoints)
 
-    self.destruct()
+    initialized = true
+    self.clearButtons()
+    self.locked = true
+    self.interactable = false
+    local position = self.getPosition()
+    position.y = -2
+    self.setPosition(position)
+
+    updatePresence()
+    Wait.time(updatePresence, 1, -1)
+end
+
+local function presenceOnIsland(color)
+    local count = 0
+    for _, presence in pairs(getObjectsWithTag("Presence")) do
+        if presence.getName() == color .. "'s Presence" then
+            local hits = upCast(presence, 1, yHeight - presence.getPosition().y)
```

Comment:
this should also be a ray cast, and should probably be changing the direction to go down instead of up

Reply:
I don't think this should be a ray cast. If the presence is hanging off the edge a bit, or if it is lined up just right between two boards, it will think it isn't on the island.